File: checkSolvability.py
RIGHT = 0
DOWN = 1
LEFT = 2
UP = 3


def createOneDimPuzzle(array, n):
    t, b, l, r = 0, n - 1, 0, n - 1
    direction = RIGHT
    newArray = []
    while t <= b and l <= r:
        if direction is RIGHT:
            i = l
            while i <= r:
                newArray.append(array[t][i])
                i += 1
            t += 1
        elif direction is DOWN:
            i = t
            while i <= b:
                newArray.append(array[i][r])
                i += 1
            r -= 1
        elif direction is LEFT:
            i = r
            while i >= l:
                newArray.append(array[b][i])
                i -= 1
            b -= 1
        elif direction is UP:
            i = b
            while i >= t:
                newArray.append(array[i][l])
                i -= 1
            l += 1
        direction = (direction + 1) % 4
    return newArray


def findPosIntab(array, value):
    count = 0
    for i in array:
        if i == value:
            return count
        count += 1
    return count


def getPermuts(array):
    count = 0
    for i in range(len(array)):
        for j in range(i + 1, len(array)):
            if array[i] > 0 and 0 < array[j] < array[i]:
                count += 1
    return count


def checkSolvability(node, size):
    array = createOneDimPuzzle(node.state, size)
    nbPermut = getPermuts(array)
    if size % 2 != 0 and nbPermut % 2 == 0:
        return True
    else:
        coord = findPosIntab(array, 0)
        if nbPermut % 2 == 0 and coord % 2 or nbPermut % 2 and coord % 2 == 0:
            return True
    return False

File: cpp/src/checkSolvability.cpp
#include "npuzzle.h"

enum directions
{
    RIGHT,
    DOWN,
    LEFT,
    UP,
};


std::vector<int> createOneDim(std::vector<int> &puzzle, int &size)
{
    int i = 0, y = 0;
    int span = size;
    int count = 0;
    std::vector<int> newArray;
    directions direction = RIGHT; 

    while (count != (size * size))
    {
        if (direction == RIGHT)
        {
            while (i < span)
            {
                newArray.push_back(puzzle[y * size + i]);
                i++;
                count++;
            }
            y++;
            i--;
        }
        else if (direction == DOWN)
        {
            while (y < span)
            {
                newArray.push_back(puzzle[y * size + i]);
                y++;
                count++;
            }
            y--;
            i--;
        }
        else if (direction == LEFT)
        {
            while (i >= size - span )
            {
                newArray.push_back(puzzle[y * size + i]);
                i--;
                count++;
            }
            i++;
            y--;
        }
        else if (direction == UP)
        {
            span--;
            while (y >= size - span)
            {
                newArray.push_back(puzzle[y * size + i]);
                y--;
                count++;
            }
            y++;
            i++;
        }
        direction = static_cast<directions>((direction + 1) % 4);
    }
    return newArray;
}

int findCoord(std::vector<int> puzzle, int value) {
    auto idx = std::find(puzzle.begin(), puzzle.end(), value) - puzzle.begin();
    return idx;
}

int getNbPermut(std::vector<int> puzzle) {
    int count = 0;
    for (size_t i = 0; i < puzzle.size(); i++) {
        for (size_t j = i + 1; j < puzzle.size(); j++) {
            if ((puzzle[i] != 0 && puzzle[j] != 0) && (puzzle[i] > puzzle[j]))
                count++;
        }
    }
    return count;
}

bool checkSolvability(std::vector<int> puzzle, int size) {
    std::vector<int> newArray = createOneDim(puzzle, size);
    int nbPermut = getNbPermut(newArray);
    int idxEmpty = findCoord(newArray, 0);
    if (size % 2 != 0)
    {   
        if (nbPermut % 2 == 0) {
            return true;
        }
    }
    else
    {
        if ((nbPermut % 2 == 0 && idxEmpty % 2) || (nbPermut % 2 && idxEmpty % 2 == 0)) {
            return true;
        }
    }
    std::cout << "not solvable, exit\n";
    return false;
}